### ISSUE TITLE
Quaternion's constructor does not recognize a 4-element numpy array

### ISSUE THREAD

**Issue by gabrielebndn** [stack-of-tasks/eigenpy#220]
The following test:
```py
import numpy as np
import eigenpy as eig

v = np.array([0.5, 0.5, -0.5, 0.5])
q = eig.Quaternion(v)
```
fails with the following error:
```
Traceback (most recent call last):
  File "test_eigenpy.py", line 5, in <module>
    q = eig.Quaternion(v)
RuntimeError: The number of rows does not fit with the matrix type.
```

On the other hand, the slightly modified test below, using a two-dimensional column array, passes without problems:
```py
import numpy as np
import eigenpy as eig

v = np.array([[0.5], [0.5], [-0.5], [0.5]])
q = eig.Quaternion(v)
```

**Comment by gabrielebndn** [comment]
Tested on: Ubuntu 20.04, Python 3.8.5, IPython 7.13.0, numpy 1.19.1

**Comment by gabrielebndn** [comment]
I couldn't manage to understand where the problem is, because it seems 3-element arrays work fine:
```py
In [6]: eig.Quaternion(np.array([1., 2., 3.]), np.array([4., 5., 6.]))                                                                  
Out[6]: (x,y,z,w) = -0.0459784  0.0919568 -0.0459784   0.993638
```

**Comment by gabrielebndn** [comment]
I think it could be easily be fixed by using `VectorXd` as an input and checking for the size, but it looks like a deeper problem with `Vector4` itself

**Comment by gabrielebndn** [comment]
Notice the constructor from a single 4-element `numpy.array` isn't checked in the unit tests

**Comment by gabrielebndn** [comment]
I think I am starting to understand the problem: in C++ there exists a single constructor accepting `MatrixBase` as input, which 
> Constructs and initializes a quaternion from either:
>     a rotation matrix expression,
>     a 4D vector expression representing quaternion coefficients.

https://eigen.tuxfamily.org/dox/classEigen_1_1Quaternion.html#aba694e0bbafcf5554b77cfe8ce69c14a

and the array gets treated like a matrix. If I do not expose the constructor from a 3D matrix, the array one works

**Comment by gabrielebndn** [comment]
I've tried working on it with no satisfactory results, I'll leave it to you. I think my comment above is a good start

**Comment by jcarpent** [comment]
Who is the you that you mention in the previous message?

**Comment by gabrielebndn** [comment]
> Who is the you that you mention in the previous message?

Well, anybody willing to work on this. I've done my best trying to fix it, sorry, but I am really not sure how to go forward, and also I think the solution may involve design choices which should probably be taken by someone more directly involved in Eigenpy's development

As for myself, I've worked quite a bit in order to identify the problem as narrowly as possible, but I don't have much time for going on with it. There are of course simple ways to go around this problem in my own code, so I think that's how I will deal with this problem for what strictly concerns me. But this is an unexpected bug, I thought you might be interested

**Comment by jcarpent** [comment]
The issue is due to Boost.Python which does some mix between equivalent constructor (as done in fact by Eigen).
I've switched the order of exposing constructor and it does work well.

**Comment by jcarpent** [comment]
Solved by #221.
